File: src/main/java/javaf/pws/secondPW/FourthTask.java
package javaf.pws.secondPW;

import java.util.NoSuchElementException;
import java.util.Scanner;

public class FourthTask {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        try{
            System.out.println("Enter something");
            String string = scanner.nextLine();
            ISelection selection = (l) -> ((l % 2) == 0);
            IMaxOrMin condition = (l, m) -> l > m;
            if (!string.isEmpty() || !string.isBlank()) System.out.println((forTask(string, condition, selection)));
            else System.out.println("Error: The string is empty or blank");
        } catch (NoSuchElementException ex){
            System.err.println("Error: String is null");
        }
    }
    private static char forTask(String string, IMaxOrMin condition, ISelection select){
        char symbol = ' ';
        int length = 0;
        int maxLength = 0;
        for (int i = 0; i < string.length(); i++){
            if ((string.charAt(i) == ' ')){
                if (select.selection(length)) {
                    maxLength = length;
                    symbol = string.charAt(i - 1);
                    length = 0;
                    break;
                }
            } else length++;
        }
        for (int i = 0; i < string.length(); i++){
            if ((string.charAt(i) == ' ')){
                if (select.selection(length) && condition.selectionMaxOrMin(length, maxLength)) {
                    maxLength = length;
                    symbol = string.charAt(i - 1);
                }
                length = 0;
            } else length++;
        }
        if (condition.selectionMaxOrMin(length, maxLength) && select.selection(length)) symbol = string.charAt(string.length() - 1);
        return symbol;
    }
    private interface ISelection { boolean selection(int n); }
    interface IMaxOrMin { boolean selectionMaxOrMin(int l, int m); }
}

File: src/main/java/kotlinf/pws/secondPW/fourthTask.kt
package kotlinf.pws.secondPW

/*В задании No3 первой практической работы реализуйте следующее: выделите основной алгоритм в отдельную функцию, куда в качестве аргумента передавайте lambda, имеющую смысл – способ сравнения двух чисел;
изменяя данную lambda, пользователь вашей функции должен иметь возможность получить информацию либо о самых длинных, либо о самых коротких словах.
Аналогичным образом передавайте в вашу функцию lambda, которая будет определять условие отбора слов.*/
fun fourthTask(string: String, condition1: (Int, Int) -> Boolean, condition2: (Int) -> Boolean): Char{
    var symbol = ' '
    var length = 0
    var maxLength = 0
    for (i in string.indices){
        if ((string[i] == ' ')){
            if (condition2(length)) {
                maxLength = length
                symbol = string[i - 1]
                break
            }
        } else length++
    }
    length = 0
    for (i in string.indices){
        if ((string[i] == ' ')){
            if (condition2(length) && condition1(length, maxLength)) {
                maxLength = length
                symbol = string[i - 1]
            }
            length = 0
        } else length++
    }
    if (condition1(length, maxLength) && condition2(length)) symbol = string[string.lastIndex]
    return symbol
}
fun main() {
    val string = readlnOrNull()
    if (string != null) println(fourthTask(string, { l: Int, maxL: Int -> l > maxL }, { l: Int -> (l % 2) == 0}))
    else println("Error")
}